File: phone-exchange-2-nodejs/SOAP/pe2_response.js
export class ExchangeInfo {
    constructor(data = {}) {
        this.PhoneNumber = data.PhoneNumber || null;
        this.Name = data.Name || null;
        this.City = data.City || null;
        this.State = data.State || null;
        this.Country = data.Country || null;
        this.LineType = data.LineType || null;
        this.TimeZone = data.TimeZone || null;
        this.Latitude = data.Latitude || null;
        this.Longitude = data.Longitude || null;
        this.SMSAddress = data.SMSAddress || null;
        this.MMSAddress = data.MMSAddress || null;
        this.PortedInfo = data.PortedInfo || null;
        this.NoteCodes = data.NoteCodes || null;
        this.NoteDescriptions = data.NoteDescriptions || null;
    }

    toString() {
        return `ExchangeInfo: PhoneNumber = ${this.PhoneNumber}, Name = ${this.Name}, City = ${this.City}, State = ${this.State}, Country = ${this.Country}, LineType = ${this.LineType}, TimeZone = ${this.TimeZone}, Latitude = ${this.Latitude}, Longitude = ${this.Longitude}, SMSAddress = ${this.SMSAddress}, MMSAddress = ${this.MMSAddress}, PortedInfo = ${this.PortedInfo}, NoteCodes = ${this.NoteCodes}, NoteDescriptions = ${this.NoteDescriptions}`;
    }
}

/**
 * Ported information for a phone number.
 */
export class PortedInfo {
    constructor(data = {}) {
        this.OriginalName = data.OriginalName || null;
        this.OriginalLineType = data.OriginalLineType || null;
        this.PortedDate = data.PortedDate || null;
        this.LATA = data.LATA || null;
    }

    toString() {
        return `PortedInfo: OriginalName = ${this.OriginalName}, OriginalLineType = ${this.OriginalLineType}, PortedDate = ${this.PortedDate}, LATA = ${this.LATA}`;
    }
}

/**
 * International phone exchange information for a validated phone number.
 */
export class InternationalExchangeInfo {
    constructor(data = {}) {
        this.PhoneNumberIn = data.NumberIn || null;
        this.CountryCode = data.CountryCode || null;
        this.FormatNational = data.FormatNational || null;
        this.Extension = data.Extension || null;
        this.Locality = data.Locality || null;
        this.LocalityMatchLevel = data.LocalityMatchLevel || null;
        this.TimeZone = data.TimeZone || null;
        this.Latitude = data.Latitude || null;
        this.Longitude = data.Longitude || null;
        this.Country = data.Country || null;
        this.CountryISO2 = data.CountryISO2 || null;
        this.CountryISO3 = data.CountryISO3 || null;
        this.FormatInternational = data.FormatInternational || null;
        this.FormatE164 = data.FormatE164 || null;
        this.Carrier = data.Carrier || null;
        this.LineType = data.LineType || null;
        this.SMSAddress = data.SMSAddress || null;
        this.MMSAddress = data.MMSAddress || null;
        this.IsValid = data.IsValid !== undefined ? data.IsValid : null;
        this.IsValidForRegion = data.IsValidForRegion !== undefined ? data.IsValidForRegion : null;
        this.NoteCodes = data.NoteCodes || null;
        this.NoteDescriptions = data.NoteDescriptions || null;
    }

    toString() {
        return `InternationalExchangeInfo: PhoneNumberIn = ${this.PhoneNumberIn}, CountryCode = ${this.CountryCode}, FormatNational = ${this.FormatNational}, Extension = ${this.Extension}, Locality = ${this.Locality}, LocalityMatchLevel = ${this.LocalityMatchLevel}, TimeZone = ${this.TimeZone}, Latitude = ${this.Latitude}, Longitude = ${this.Longitude}, Country = ${this.Country}, CountryISO2 = ${this.CountryISO2}, CountryISO3 = ${this.CountryISO3}, FormatInternational = ${this.FormatInternational}, FormatE164 = ${this.FormatE164}, Carrier = ${this.Carrier}, LineType = ${this.LineType}, SMSAddress = ${this.SMSAddress}, MMSAddress = ${this.MMSAddress}, IsValid = ${this.IsValid}, IsValidForRegion = ${this.IsValidForRegion}, NoteCodes = ${this.NoteCodes}, NoteDescriptions = ${this.NoteDescriptions}`;
    }
}

/**
 * Error object for PE2 API responses.
 */
export class Error {
    constructor(data = {}) {
        this.Desc = data.Desc || null;
        this.TypeCode = data.TypeCode || null;
        this.DescCode = data.DescCode || null;
        this.Type = data.Type || null;
    }

    toString() {
        return `Error: Desc = ${this.Desc}, TypeCode = ${this.TypeCode}, DescCode = ${this.DescCode}, Type = ${this.Type}`;
    }
}

/**
 * Response from PE2 GetExchangeInfo and GetInternationalExchangeInfo APIs, containing phone exchange information.
 */
export class PE2Response {
    constructor(data = {}) {
        const exchangeInfo = Array.isArray(data?.ExchangeInfoResults?.ExchangeInfo)
            ? data.ExchangeInfoResults.ExchangeInfo
            : [];

        this.ExchangeInfo = exchangeInfo.map(info => new ExchangeInfo(info));
        this.ExchangeInfoResults = this.ExchangeInfo;

        this.InternationalExchangeInfo = data.InternationalExchangeInfo

        this.Error = data?.Error ? new Error(data.Error) : null;
    }

    toString() {
        const exchangeInfoString = this.ExchangeInfoResults.length
            ? this.ExchangeInfoResults.map(info => info.toString()).join('; ')
            : 'null';

        const internationalInfoString = this.InternationalExchangeInfo.length
            ? this.InternationalExchangeInfo.map(info => info.toString()).join('; ')
            : 'null';

        return `PE2Response: ExchangeInfoResults = [${exchangeInfoString}], InternationalExchangeInfo = [${internationalInfoString}], Error = ${this.Error ? this.Error.toString() : 'null'}`;
    }
}

export default PE2Response;

File: phone-exchange-2-dot-net/REST/PE2Response.cs
using System.Runtime.Serialization;
using System.Linq;

namespace phone_exchange_2_dot_net.REST
{
    /// <summary>
    /// Response from PE2 GetExchangeInfo and GetInternationalExchangeInfo APIs, containing phone exchange information.
    /// </summary>
    [DataContract]
    public class PE2Response
    {
        public ExchangeInfo[] ExchangeInfoResults { get; set; }
        public InternationalExchangeInfo InternationalExchangeInfo { get; set; }
        public Error Error { get; set; }

        public override string ToString()
        {
            string exchangeInfoStr = ExchangeInfoResults != null
                ? string.Join("\n", ExchangeInfoResults.Select(r => r.ToString()))
                : "null";
            string internationalInfoStr = InternationalExchangeInfo != null
                ? InternationalExchangeInfo.ToString()
                : "";
            return $"PE2Response:\n" +
                   $"ExchangeInfoResults:\n{exchangeInfoStr}\n" +
                   $"InternationalExchangeInfo:\n{internationalInfoStr}\n" +
                   $"Error: {(Error != null ? Error.ToString() : "null")}";
        }
    }

    /// <summary>
    /// Phone exchange information for a validated phone number (USA/Canada).
    /// </summary>
    [DataContract]
    public class ExchangeInfo
    {
        public string PhoneNumber { get; set; }
        public string Name { get; set; }
        public string City { get; set; }
        public string State { get; set; }
        public string Country { get; set; }
        public string LineType { get; set; }
        public string TimeZone { get; set; }
        public string Latitude { get; set; }
        public string Longitude { get; set; }
        public string SMSAddress { get; set; }
        public string MMSAddress { get; set; }
        public PortedInfo PortedInfo { get; set; }
        public string NoteCodes { get; set; }
        public string NoteDescriptions { get; set; }

        public override string ToString()
        {
            string portedInfoStr = PortedInfo != null
                ? PortedInfo.ToString()
                : "";
            return $"ExchangeInfo:\n" +
                   $"PhoneNumber: {PhoneNumber}\n" +
                   $"Name: {Name}\n" +
                   $"City: {City}\n" +
                   $"State: {State}\n" +
                   $"Country: {Country}\n" +
                   $"LineType: {LineType}\n" +
                   $"TimeZone: {TimeZone}\n" +
                   $"Latitude: {Latitude}\n" +
                   $"Longitude: {Longitude}\n" +
                   $"SMSAddress: {SMSAddress}\n" +
                   $"MMSAddress: {MMSAddress}\n" +
                   $"PortedInfo: {portedInfoStr}\n" +
                   $"NoteCodes: {NoteCodes}\n" +
                   $"NoteDescriptions: {NoteDescriptions}";
        }
    }

    /// <summary>
    /// Ported information for a phone number.
    /// </summary>
    [DataContract]
    public class PortedInfo
    {
        public string OriginalName { get; set; }
        public string OriginalLineType { get; set; }
        public string PortedDate { get; set; }
        public string LATA { get; set; }

        public override string ToString()
        {
            return $"OriginalName: {OriginalName}, OriginalLineType: {OriginalLineType}, PortedDate: {PortedDate}, LATA: {LATA}";
        }
    }

    /// <summary>
    /// International phone exchange information for a validated phone number.
    /// </summary>
    [DataContract]
    public class InternationalExchangeInfo
    {
        public string PhoneNumberIn { get; set; }
        public string CountryCode { get; set; }
        public string FormatNational { get; set; }
        public string Extension { get; set; }
        public string Locality { get; set; }
        public string LocalityMatchLevel { get; set; }
        public string TimeZone { get; set; }
        public string Latitude { get; set; }
        public string Longitude { get; set; }
        public string Country { get; set; }
        public string CountryISO2 { get; set; }
        public string CountryISO3 { get; set; }
        public string FormatInternational { get; set; }
        public string FormatE164 { get; set; }
        public string Carrier { get; set; }
        public string LineType { get; set; }
        public string SMSAddress { get; set; }
        public string MMSAddress { get; set; }
        public bool IsValid { get; set; }
        public bool IsValidForRegion { get; set; }
        public string NoteCodes { get; set; }
        public string NoteDescriptions { get; set; }

        public override string ToString()
        {
            return $"InternationalExchangeInfo:\n" +
                   $"PhoneNumberIn: {PhoneNumberIn}\n" +
                   $"CountryCode: {CountryCode}\n" +
                   $"FormatNational: {FormatNational}\n" +
                   $"Extension: {Extension}\n" +
                   $"Locality: {Locality}\n" +
                   $"LocalityMatchLevel: {LocalityMatchLevel}\n" +
                   $"TimeZone: {TimeZone}\n" +
                   $"Latitude: {Latitude}\n" +
                   $"Longitude: {Longitude}\n" +
                   $"Country: {Country}\n" +
                   $"CountryISO2: {CountryISO2}\n" +
                   $"CountryISO3: {CountryISO3}\n" +
                   $"FormatInternational: {FormatInternational}\n" +
                   $"FormatE164: {FormatE164}\n" +
                   $"Carrier: {Carrier}\n" +
                   $"LineType: {LineType}\n" +
                   $"SMSAddress: {SMSAddress}\n" +
                   $"MMSAddress: {MMSAddress}\n" +
                   $"IsValid: {IsValid}\n" +
                   $"IsValidForRegion: {IsValidForRegion}\n" +
                   $"NoteCodes: {NoteCodes}\n" +
                   $"NoteDescriptions: {NoteDescriptions}";
        }
    }

    /// <summary>
    /// Error object for PE2 API responses.
    /// </summary>
    [DataContract]
    public class Error
    {
        public string Type { get; set; }
        public string Desc { get; set; }
        public string TypeCode { get; set; }
        public string DescCode { get; set; }

        public override string ToString()
        {
            return $"Desc: {Desc}, TypeCode: {TypeCode}, DescCode: {DescCode}, Type: {Type}";
        }
    }
}